helm & e2e: add option to enable read affinity for rbd

iPraveenParihar opened the pull request:
<!-- Please take a look at our [Contributing](https://github.com/ceph/ceph-csi/blob/devel/docs/development-guide.md#Code-contribution-workflow)
documentation before submitting a Pull Request!
Thank you for contributing to ceph-csi! -->

# Describe what this PR does #

- helm: add option to enable read affinity for rbd
This commit adds --enable-read-affinity flag to
enable read affinity for rbd

- e2e: added test to verify read affinity functionality
e2e test case is added to test if read affinity is enabled by 
verifying `read_from_replica=localize` option is passed.

### Closes: #3642 
---

iPraveenParihar committed:
helm: add option to enable read affinity for rbd

This commit adds --enable-read-affinity flag to
enable read affinity for rbd

Signed-off-by: Praveen M <m.praveen@ibm.com>

Rakshith-R commented on charts/ceph-csi-rbd/values.yaml:
```diff
@@ -279,6 +279,23 @@ topology:
+  crushLocationLabels:
+    - topology.kubernetes.io/region
+    - topology.kubernetes.io/zone
+    - topology.rook.io/datacenter
+    - topology.rook.io/room
+    - topology.rook.io/pod
+    - topology.rook.io/pdu
+    - topology.rook.io/row
+    - topology.rook.io/rack
+    - topology.rook.io/chassis
```
We can just have the above labels as examples.
These rook specific labels can be removed.

Madhu-1 commented on charts/ceph-csi-rbd/templates/nodeplugin-daemonset.yaml:
```diff
@@ -86,6 +86,10 @@ spec:
 {{- if .Values.nodeplugin.profiling.enabled }}
             - "--enableprofiling={{ .Values.nodeplugin.profiling.enabled }}"
+{{- end }}
+{{- if .Values.readAffinity.enabled }}
+            - "--enable-read-affinity=true"
```
This will be problematic if we choose to enable it by default later for some reason and the user should be able to disable it

iPraveenParihar replied:
@Madhu-1,  I didn't get you on this. Can you explain bit more?

Madhu-1 replied:
```suggestion
            - "--enable-read-affinity={{  .Values.readAffinity.enabled }}"
 {{- if .Values.readAffinity.enabled }}
            - "--crush-location-labels={{ .Values.readAffinity.crushLocationLabels | join "," }}"
{{- end }}
```

The above change will allow users to disable the ReadAffinity if its get enabled by default.

iPraveenParihar replied:
Okay, I got it
Thanks!

Madhu-1 commented on charts/ceph-csi-rbd/README.md:
```diff
@@ -176,6 +176,8 @@ charts and their default values.
 | `topology.enabled`                             | Specifies whether topology based provisioning support should be exposed by CSI                                                                       | `false`                                            |
 | `topology.domainLabels`                        | DomainLabels define which node labels to use as domains for CSI nodeplugins to advertise their domains                                               | `{}`                                               |
+| `readAffinity.enabled` | Enable read affinity for RBD volumes. Recommended to set to true if running kernel 5.8 or newer. | `false` |
+| `readAffinity.crushLocationLabels` | Define which node labels to use as CRUSH location. This should correspond to the values set in the CRUSH map. | labels listed [here](https://github.com/rook/rook/blob/master/Documentation/CRDs/Cluster/ceph-cluster-crd.md#osd-topology) |
```
Can you point to the ceph doc instead of the Rook doc for this one? if not, this can never be used with regular ceph clusters?

Rakshith-R replied:
Praveen, The default is empty list `[]`, 

add a sentence "For more information, click [here]"
https://github.com/ceph/ceph-csi/blob/v3.9.0/docs/deploy-rbd.md#read-affinity-using-crush-locations-for-rbd-volumes

Users can refer to this section for more details

iPraveenParihar replied:
@Rakshith-R, I have made the changes

iPraveenParihar committed:
e2e: added test to verify read affinity functionality

e2e test case is added to test if read affinity is enabled by
verifying read_from_replica=localize option is passed

Signed-off-by: Praveen M <m.praveen@ibm.com>

iPraveenParihar commented on e2e/deployment.go:
```diff
@@ -240,6 +240,10 @@ type yamlResourceNamespaced struct {
+
+	// enable read affinity support (for RBD)
+	enableReadAffinity  bool
+	crushLocationLabels string
```
Hi @Madhu-1,
golangci-lint is faling because of maligned struct

```
e2e/deployment.go:233:29: struct of size 80 bytes could be of size 72 bytes:
struct{
	filename           	string,
	namespace          	string,
	domainLabel        	string,
	crushLocationLabels	string,
	oneReplica         	bool,
	enableTopology     	bool,
	enableReadAffinity 	bool,
}
``` (maligned)
type yamlResourceNamespaced struct {
```

Do we reorder as suggested?

Madhu-1 replied:
yes it's best to do it as suggested.

iPraveenParihar commented:
/test ci/centos/mini-e2e/k8s-1.26

iPraveenParihar commented:
/test ci/centos/mini-e2e-helm/k8s-1.26

iPraveenParihar commented:
/test ci/centos/mini-e2e-helm/k8s-1.27

iPraveenParihar commented:
/test ci/centos/mini-e2e/k8s-1.27

iPraveenParihar commented:
/test ci/centos/mini-e2e/k8s-1.28

iPraveenParihar commented:
/test ci/centos/mini-e2e-helm/k8s-1.28

iPraveenParihar commented:
@Rakshith-R,
For the e2e test case - verifying just the presence of the `read_from_replica=localize` option is enough?
As the crush location label:value will be in random order.

example - `crush_location=zone:east-zone1|region:east` or it may be `crush_location=region:east|zone:east-zone1`.
In this case, I'll have to parse the`crush_location` value and then verify each label:value individually

Rakshith-R commented:
> @Rakshith-R, For the e2e test case - verifying just the presence of the `read_from_replica=localize` option is enough? As the crush location label:value will be in random order.
> 
> example - `crush_location=zone:east-zone1|region:east` or it may be `crush_location=region:east|zone:east-zone1`. In this case, I'll have to parse the`crush_location` value and then verify each label:value individually

It'll be great to parse and verify the key value pairs too.

iPraveenParihar commented:
/test ci/centos/mini-e2e/k8s-1.26

iPraveenParihar commented:
/test ci/centos/mini-e2e/k8s-1.27

iPraveenParihar commented:
/test ci/centos/mini-e2e/k8s-1.28

iPraveenParihar commented:
/test ci/centos/mini-e2e/k8s-1.28

iPraveenParihar commented:
/test ci/centos/mini-e2e-helm/k8s-1.26

iPraveenParihar commented:
/test ci/centos/mini-e2e-helm/k8s-1.27

iPraveenParihar commented:
/test ci/centos/mini-e2e-helm/k8s-1.28

Madhu-1 commented on charts/ceph-csi-rbd/values.yaml:
```diff
@@ -279,6 +279,17 @@ topology:
+  # NOTE: the value here serves as an example
+  crushLocationLabels:
+    - topology.kubernetes.io/region
+    - topology.kubernetes.io/zone
```
This needs to be commented out and set in install-helm.sh script

iPraveenParihar commented:
/test ci/centos/mini-e2e-helm/k8s-1.26

iPraveenParihar commented:
/test ci/centos/mini-e2e-helm/k8s-1.27

iPraveenParihar commented:
/test ci/centos/mini-e2e-helm/k8s-1.28

Madhu-1 commented:
@iPraveenParihar as your changes don't have any impact from the kubernetes version, just run one test to ensure its working as expected.

iPraveenParihar commented:
> @iPraveenParihar as your changes don't have any impact from the kubernetes version, just run one test to ensure its working as expected.

😅 oh okay, got it.

Madhu-1 commented on e2e/pod.go:
```diff
@@ -623,3 +624,83 @@ func verifySeLinuxMountOption(
+		return fmt.Errorf("failed to get pod logs from container %s/%s/%s : %w", ns, nodepluginPodName, cn, err)
+	}
+
+	if !strings.Contains(logs, readFromReplicaOption) {
+		return fmt.Errorf("option %s not found in logs: %s", readFromReplicaOption, logs)
+	}
```
IMO this is not the best solution because if we log this option anywhere but we don't pass it to map/mount option this test will pass. For RBD how about checking the map options in the device config file? like `/sys/devices/rbd/0/config_info` this ensures that map option is applied for the mapped rbd image.

iPraveenParihar replied:
@Madhu-1,
looks like I can’t read `/sys/devices/rbd/0/config_info`  file
```
root@csi-rbd-demo-pod:/sys/devices/rbd/0# ls -l config_info 
-r-------- 1 root root 4096 Sep 21 05:35 config_info
root@csi-rbd-demo-pod:/sys/devices/rbd/0# cat config_info 
cat: config_info: Operation not permitted
root@csi-rbd-demo-pod:/sys/devices/rbd/0# 
```

Madhu-1 replied:
@iPraveenParihar not from the demo pod but check it from the csi-rbdplugin container

iPraveenParihar replied:
I got it, I was doing exec in the demo pod.

From csi-rbdplugin-xxx pod csi-rbdplugin container
```
[root@m1 0]# cat config_info 
10.99.51.241:6789 name=admin,key=client.admin,crush_location=region:east|zone:east-zone1,read_from_replica=localize replicapool csi-vol-134dadfd-4af1-4026-b79f-8fd265974a2b -
```
Thanks

Madhu-1 commented on e2e/rbd.go:
```diff
@@ -275,6 +283,14 @@ var _ = Describe("RBD", func() {
 			if err != nil {
 				framework.Failf("failed to create node label: %v", err)
 			}
+			err = createNodeLabel(f, crushLocationRegionLabel, crushLocationRegionValue)
+			if err != nil {
+				framework.Failf("failed to create node label: %v", err)
+			}
+			err = createNodeLabel(f, crushLocationZoneLabel, crushLocationZoneValue)
+			if err != nil {
+				framework.Failf("failed to create node label: %v", err)
+			}
```
lets enhance CreateNodeLable to take map[string]string and set the label once (not as part of this PR)

iPraveenParihar replied:
Created a issue - #4146

Madhu-1 commented on e2e/rbd.go:
```diff
@@ -409,6 +425,15 @@ var _ = Describe("RBD", func() {
 		if err != nil {
 			framework.Failf("failed to delete node label: %v", err)
 		}
+		// Remove the CRUSH Location labels
```
same here as well, lets update deleteNodeLabel to take multiple labels (if possible)

iPraveenParihar commented:
/test ci/centos/mini-e2e/k8s-1.28

Madhu-1 commented on e2e/pod.go:
```diff
@@ -623,3 +624,82 @@ func verifySeLinuxMountOption(
+	opt := metav1.ListOptions{
+		LabelSelector: selector,
+	}
+	command := "cat /sys/devices/rbd/0/config_info"
```
we cant have `0` as hardcoded value, we need to get the device number from mapped image.

iPraveenParihar replied:
Okay, I thought during each e2e test case we will be having only one device so will use `0`.

iPraveenParihar replied:
@Madhu-1
```
bash-4.4$ rbd device list --format=json
[{"id":"0","pool":"replicapool","namespace":"","name":"csi-vol-37f39fc8-e26d-40f3-a297-acd16e7afbb8","snap":"-","device":"/dev/rbd0"}]
``` 

does this `id` refer to device number?

Madhu-1 commented on charts/ceph-csi-rbd/values.yaml:
```diff
@@ -279,6 +279,17 @@ topology:
+  # Define which node labels to use as CRUSH location.
+  # This should correspond to the values set in the CRUSH map.
+  # NOTE: the value here serves as an example
+  crushLocationLabels:
```
comment out the key as well and even enabled key as well.

Madhu-1 reviewed (commented):
small nit

iPraveenParihar commented:
/test ci/centos/mini-e2e/k8s-1.28

iPraveenParihar commented:
/test ci/centos/mini-e2e/k8s-1.28

Madhu-1 commented on e2e/pod.go:
```diff
@@ -623,3 +624,102 @@ func verifySeLinuxMountOption(
+	}
+
+	if configInfo == "" {
+		return fmt.Errorf("failed to get config_info file")
```
use errors.New() when error formatting is not required

Madhu-1 reviewed (commented):
one last nit, LGTM

Madhu-1 commented:
@iPraveenParihar linter is still failing

iPraveenParihar commented:
> @iPraveenParihar linter is still failing

yea, checking it -   `283:3     warning  comment not indented like content  (comments-indentation)`
But not able to figure it out

```
readAffinity:
  # Enable read affinity for RBD volumes. Recommended to
  # set to true if running kernel 5.8 or newer.
  # enabled: false
  # Define which node labels to use as CRUSH location.
  # This should correspond to the values set in the CRUSH map.
  # NOTE: the value here serves as an example
  # crushLocationLabels:
  #  - topology.kubernetes.io/region
  #  - topology.kubernetes.io/zone
```

If I uncomment the `enabled` and `crushLocationLabels` key, lint passes

iPraveenParihar commented:
@Madhu-1 , commenting `enabled` key throws lint error
```
Error:  templates/: template: ceph-csi-rbd/templates/nodeplugin-daemonset.yaml:90:48: executing "ceph-csi-rbd/templates/nodeplugin-daemonset.yaml" at <.Values.readAffinity.enabled>: nil pointer evaluating interface {}.enabled
```

we can just comment out the `crushLocationLabels`?

Madhu-1 commented:
> @Madhu-1 , commenting `enabled` key throws lint error
> 
> ```
> Error:  templates/: template: ceph-csi-rbd/templates/nodeplugin-daemonset.yaml:90:48: executing "ceph-csi-rbd/templates/nodeplugin-daemonset.yaml" at <.Values.readAffinity.enabled>: nil pointer evaluating interface {}.enabled
> ```
> 
> we can just comment out the `crushLocationLabels`?

you can add one more check to ensure that `Values.readAffinity` is set?

iPraveenParihar commented:
/test ci/centos/mini-e2e-helm/k8s-1.28

Rakshith-R commented:
@iPraveenParihar you can use https://docs.mergify.com/commands/rebase to rebase, It'll preserve the approvals.

iPraveenParihar commented:
@Mergifyio queue

Rakshith-R commented:
@mergifyio queue
